packages/flutter_tools/lib/src/android/gradle_errors.dart

```diff
@@ -234,18 +229,71 @@ final GradleHandledError networkErrorHandler = GradleHandledError(
       '${globals.logger.terminal.warningMark} '
       'Gradle threw an error while downloading artifacts from the network.'
     );
-    try {
+    return GradleBuildStatus.retry;
+  },
+  eventLabel: 'network',
+);
+
+/// Handles corrupted jar or other types of zip files.
+///
+/// If a terminal is attached, this handler prompts the user if they would like to
+/// delete the $HOME/.gradle directory prior to retrying the build.
+///
+/// If this handler runs on a bot (e.g. a CI bot), the $HOME/.gradle is automatically deleted.
+///
+/// See also:
+///  * https://github.com/flutter/flutter/issues/51195
+///  * https://github.com/flutter/flutter/issues/89959
+///  * https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home
+@visibleForTesting
+final GradleHandledError zipExceptionHandler = GradleHandledError(
+  test: _lineMatcher(const <String>[
+    'java.util.zip.ZipException: error in opening zip file',
+  ]),
+  handler: ({
+    required String line,
+    required FlutterProject project,
+    required bool usesAndroidX,
+    required bool multidexEnabled,
+  }) async {
+    globals.printError(
+      '${globals.logger.terminal.warningMark} '
+      'Your .gradle directory under the home directory might be corrupted.'
+    );
+    bool shouldDeleteUserGradle = await globals.botDetector.isRunningOnBot;
+    if (!shouldDeleteUserGradle && globals.terminal.stdinHasTerminal) {
+      try {
+        final String selection = await globals.terminal.promptForCharInput(
+          <String>['y', 'n'],
+          logger: globals.logger,
+          prompt: 'Do you want to delete the .gradle directory under the home directory?',
+          defaultChoiceIndex: 0,
+        );
+        shouldDeleteUserGradle = selection == 'y';
+      } on StateError catch(e) {
+        globals.printError(
+          e.message,
+          indent: 0,
+        );
+      }
+    }
+    if (shouldDeleteUserGradle) {
       final String? homeDir = globals.platform.environment['HOME'];
-      if (homeDir != null) {
-        final Directory directory = globals.fs.directory(globals.fs.path.join(homeDir, '.gradle'));
-        ErrorHandlingFileSystem.deleteIfExists(directory, recursive: true);
+      if (homeDir == null) {
+        globals.logger.printStatus("Could not delete .gradle directory because there isn't a HOME env variable");
```

Comment:
Out of complete curiosity, how common would it be for someone to have moved it since the home directory is the default? If decently common, would it be reasonable to allow them provide the path to delete it?

Reply:
I don't think it is common. The user would need to do something different than what Gradle does by default: https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home

I'm also not sure if it's possible.- I'm not aware of a GRADLE_HOME variable or similar that would inform Gradle where this directory is.